### DISCUSSION TITLE
istio routing order when using multiple virtualservices

### DISCUSSION THREAD

**Discussion by sambitr** [istio/istio#53551] [Ideas]
(This is used to request new product features, please visit <https://github.com/istio/istio/discussions> for questions on using Istio)

we deploy 2 virtualservices from 2 Helm chart deployments for their own microservice, which uses host as * and shared istio-ingressgateway

```
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: demo-ingress-api
  namespace: demo
spec:
  gateways:
  - istio-system/istio-ingressgateway
  hosts:
  - '*'
  http:
  - match:
    - uri:
        prefix: /demo/api/
    route:
    - destination:
        host: api
        port:
          number: 80
    timeout: 30s
  - match:
    - uri:
        prefix: /demo/swagger
    route:
    - destination:
        host: api
        port:
          number: 80
    timeout: 30s
---
$ k get virtualservices -n demo demo-ingress-web -o yaml
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
  name: demo-ingress-web
  namespace: demo
spec:
  gateways:
  - istio-system/istio-ingressgateway
  hosts:
  - '*'
  http:
  - match:
    - uri:
        prefix: /demo/
	rewrite:
	  uri: /
    route:
    - destination:
        host: web
        port:
          number: 80
    timeout: 30s
```

they are deployed in the order in which they are defined above.

The way istio orders the route is:

```
$ istioctl proxy-config routes istio-ingressgateway-5688c54c67-g8dkx -n istio-system
NAME                                                  VHOST NAME     DOMAINS     MATCH                  VIRTUAL SERVICE
https.443.https.istio-ingressgateway.istio-system     *:443          *           /demo*                 demo-ingress-web.demo
https.443.https.istio-ingressgateway.istio-system     *:443          *           /demo/api/*            demo-ingress-api.demo
https.443.https.istio-ingressgateway.istio-system     *:443          *           /demo/swagger*         demo-ingress-api.demo
```

So what is happening is that, all the traffic that should go to api virtual service, when called: <host-name>/demo/api is going to web virtual service.

I feel the routing order is responsible for that.

Interestingly, if I keep the web virtualservice like:

```
$ k get virtualservices -n demo demo-ingress-web -o yaml
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
  name: demo-ingress-web
  namespace: demo
spec:
  gateways:
  - istio-system/istio-ingressgateway
  hosts:
  - '*'
  http:
  - match:
    - uri:
        prefix: /
    route:
    - destination:
        host: web
        port:
          number: 80
    timeout: 30s
```

routing turns out as below and it solves the problem:

```
NAME                                                  VHOST NAME     DOMAINS     MATCH                  VIRTUAL SERVICE
https.443.https.istio-ingressgateway.istio-system     *:443          *           /demo/api/*            demo-ingress-api.demo
https.443.https.istio-ingressgateway.istio-system     *:443          *           /demo/swagger*         demo-ingress-api.demo
https.443.https.istio-ingressgateway.istio-system     *:443          *           /*                     demo-ingress-web.demo
```

**What should be better, if we can control the routing order incase we deploy it from multiple deployments.**

****FYI:** I know if we keep it in a single virtualservice, that works. But keeping it in multiple virtual services help our application team understand it better, so that they can modify when they require it.

**Describe alternatives you've considered**

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Dual Stack
[ ] Installation
[ X ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

**Comment by howardjohn** [comment]
Better ordering is a part of the https://gateway-api.sigs.k8s.io/ which Istio has full support for

**Reply by sambitr** [reply]
hey @howardjohn any idea why the current way of routing is behaved like how I mentioned? why /demo is coming at the top when it is deployed at the end?

**Comment by sambitr** [comment]
hi @howardjohn  I want to understand how routing is decided. I read from the documentation that the more specific routing would be picked. In that case, /demo/api/quiz should go to  api virtualservice. But it is going to web virtual service.

istio ingressgateway pod log:

[2024-10-17T17:22:42.608Z] "GET _/demo/api/quiz/categories_ HTTP/1.1" 200 - via_upstream - "-" 0 30228 3 - "170.194.32.52, 10.122.175.37:10615,10.111.213.134" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36" "0cd51b1f-56bf-407e-8d31-99bfb0d9349a" "domain-int.northeurope.com" "172.0.2.55:8080" outbound|80||**web.demo.svc.cluster.local** 172.0.3.32:50168 172.0.3.32:8443 10.111.213.134:13419 domain-int.northeurope.com -

it seems Istio is not matching the more specific ones first? Is it because the virtualservices are deployed from 2 manifest files?

Please help me on this @howardjohn .

FYI: I am using istio: 1.21.6 and AKS: 1.29.8

**Comment by sambitr** [comment]
This was discussed in another issue: https://github.com/istio/istio/issues/53564

Routing order also gets defined by Istio, the order in which they are created
